Src/FluentAssertions/Collections/GenericCollectionAssertions.cs

```diff
@@ -944,6 +944,98 @@ public AndConstraint<TAssertions> ContainInOrder(params T[] expected)
         return new AndConstraint<TAssertions>((TAssertions)this);
     }
 
+    /// <summary>
+    /// Expects the current collection to contain the specified elements in the exact same order, and to be consecutive.
+    /// using their <see cref="object.Equals(object)" /> implementation.
+    /// </summary>
+    /// <param name="expected">An <see cref="IEnumerable{T}"/> with the expected elements.</param>
+    public AndConstraint<TAssertions> ContainInConsecutiveOrder(params T[] expected)
+    {
+        return ContainInConsecutiveOrder(expected, string.Empty);
+    }
+
+    /// <summary>
+    /// Expects the current collection to contain the specified elements in the exact same order, and to be consecutive.
+    /// </summary>
+    /// <remarks>
+    /// Elements are compared using their <see cref="object.Equals(object)" /> implementation.
+    /// </remarks>
+    /// <param name="expected">An <see cref="IEnumerable{T}"/> with the expected elements.</param>
+    /// <param name="because">
+    /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+    /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+    /// </param>
+    /// <param name="becauseArgs">
+    /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+    /// </param>
+    /// <exception cref="ArgumentNullException"><paramref name="expected"/> is <c>null</c>.</exception>
+    public AndConstraint<TAssertions> ContainInConsecutiveOrder(IEnumerable<T> expected, string because = "",
+        params object[] becauseArgs)
+    {
+        Guard.ThrowIfArgumentIsNull(expected, nameof(expected), "Cannot verify ordered containment against a <null> collection.");
+
+        bool success = Execute.Assertion
+            .BecauseOf(because, becauseArgs)
+            .ForCondition(Subject is not null)
+            .FailWith("Expected {context:collection} to contain {0} in order{reason}, but found <null>.", expected);
+
+        if (success)
+        {
+            IList<T> expectedItems = expected.ConvertOrCastToList();
```

Comment:
I'm struggling to understand this implementation. I was expecting a much simpler algorithm (preferably using separate methods), probably because it's too long. Something like:

1. Scan the subject collection until you find the first item from the expected collection
2. From that point on, verify whether all successive items match. Remember the index of the first mismatching item
3. If not, restart scanning from the last index you found in 1
4. If no full match is found, use the index found in 2 to report a proper error

Reply:
I was trying to keep the code in a similar style as the existing function.

I've rewritten the code based on the suggestion. I think that it reads better now, and found a problem in a test case, and a missing test case in the process. I hope that it's what you had in mind.

I do think that I need to take another look at the not version of the function, but as I am about to go away on vacation for a week I'll do that when I return.